File: LEETCODE/LEETCODE-JAVA/2131.java
// 2131. Longest Palindrome by Concatenating Two Letter Words
// https://leetcode.com/problems/longest-palindrome-by-concatenating-two-letter-words/

//Method-1 Using Map
class Solution {
    public int longestPalindrome(String[] words) {
        Map<String, Integer> map = new HashMap<>();
        int result = 0;

        for (String word : words) {
            String rev = new StringBuilder(word).reverse().toString();
            if (map.getOrDefault(rev, 0) > 0) {
                result += 4;
                map.put(rev, map.get(rev) - 1);
            } else {
                map.put(word, map.getOrDefault(word, 0) + 1);
            }
        }

        for (Map.Entry<String, Integer> entry : map.entrySet()) {
            String word = entry.getKey();
            int count = entry.getValue();

            if (word.charAt(0) == word.charAt(1) && count > 0) {
                result += 2;
                break;
            }
        }

        return result;
    }
}

File: LEETCODE/LEETCODE-CPP/2131.cpp
// 2131. Longest Palindrome by Concatenating Two Letter Words
// https://leetcode.com/problems/longest-palindrome-by-concatenating-two-letter-words/

// Method-1 Using Map
class Solution
{
public:
    int longestPalindrome(vector<string> &words)
    {
        int n = words.size();
        unordered_map<string, int> mp;
        int count = 0;

        for (string &x : words)
        {
            string temp = x;

            reverse(temp.begin(), temp.end());

            if (mp[temp] > 0)
            {
                count += 4;
                mp[temp]--;
            }
            else
            {
                mp[x]++;
            }
        }

        for (auto &[word, freq] : mp)
        {
            if (word[0] == word[1] && freq > 0)
            {
                count = count + (2 * freq);
                break;
            }
        }

        return count;
    }
};